packages/sdk-core/src/Framework.ts

```diff
@@ -103,7 +109,19 @@ export default class Framework {
                 : options.customSubgraphQueriesEndpoint ||
                   getSubgraphQueriesEndpoint(options);
 
-        const network = await options.provider.getNetwork();
+        const provider =
+            options.provider != null
+                ? options.provider
+                : options.web3 != null
+                ? new ethers.providers.Web3Provider(
+                      options.web3.currentProvider as
+                          | ethers.providers.ExternalProvider
+                          | ethers.providers.JsonRpcFetchFunc
+                  )
+                : // NOTE: this will not be null as we check this in our validate function
+                  options.hardhatEthers!.provider;
+
```

Comment:
This type of code is a bit smelly because cyclomatic complexity is increasing and types are not clearly communicating. `provider`, `web3` and `hardhatEthers` have a priority order, i.e. if all three are passed in then some will get ignored. Because this is an SDK and entrypoint to the framework then it's probably OK because we will have it well-tested and well-documented. But generally in this type of situation you would want to achieve code similar to this: `provider: Web3 | HardhatProvider | EthersProvider` and this case the types communicate more clearly.